Tools/Scripts/webkitpy/xcode/simulated_device.py

```diff
@@ -420,30 +436,44 @@ def initialize_devices(cls, requests, host=None, name_base='Managed', simulator_
         if not hasattr(requests, '__iter__'):
             requests = [requests]
 
+        # Parse user-specified UDIDs
+        if type(udids) is str:
+            udids = udids.split(',')
+
         # Check running sims
+        deferred_booted_devices = []
         for device in cls.available_devices(host):
-            matched_request = cls._does_fulfill_request(device, requests)
+            matched_request = cls._does_fulfill_request(device, requests, True)
             if matched_request is None:
                 continue
-            requests.remove(matched_request)
-            _log.debug(u'Attached to running simulator {}'.format(device))
-            SimulatedDeviceManager.INITIALIZED_DEVICES.append(device)
 
-            # DeviceRequests are compared by reference
-            requests_copy = [request for request in requests]
+            device_is_booted = device.platform_device.is_booted_or_booting()
+            if udids is not None and device.platform_device.udid not in udids:
+                # Only add booted devices, we'll add non-booted if we still need them later
+                if device_is_booted:
+                    deferred_booted_devices.append((matched_request, device))
+                continue
+            else:
+                cls._boot_device(device, host) if not device_is_booted else SimulatedDeviceManager.INITIALIZED_DEVICES.append(device)
 
-            # Merging requests means that if 4 devices are requested, but only one is running, these
-            # 4 requests will be fulfilled by the 1 running device.
-            for request in requests_copy:
-                if not request.merge_requests:
-                    continue
-                if not request.use_booted_simulator:
-                    continue
-                if request.device_type != device.device_type and not request.allow_incomplete_match:
-                    continue
-                if request.device_type.software_variant != device.device_type.software_variant:
-                    continue
-                requests.remove(request)
+            _log.debug(u'Attached to requested simulator {}'.format(device))
+            requests.remove(matched_request)
+            requests = cls._validate_running_device_against_requests(requests, device)
+
+        # If the user passed UDIDs, we should check for matches among remaining booted devices and
+        # then resort to using any available ones or creating new ones that match the requests.
+        if udids is not None and len(deferred_booted_devices) and len(requests):
+            for matched_request, device in deferred_booted_devices:
```

Comment:
If I understand correctly, if a device doesn't match, `matched_request = cls._does_fulfill_request(device, requests, True) = None`? And you indiscriminately add booted devices to the deferred list in line 454.

If you add a non-matching request to your deferred list and iterate over them here, do you want to be treating non-matches the same as matches? (Just checking if this behavior is intended, since I don't understand the full process.)

Reply:
Discussed offline, @briannafan didn't see this above:

```
if matched_request is None:
    continue
```

Reply:
i am blind

Reply:
My brain tends to ignore anything not red or green oops